### ISSUE TITLE
Documentation Improvements in Targeting New Users

### ISSUE THREAD

**Issue by khajavi** [zio/zio#7121]
Our goal is to teach programmers from other ecosystems (e.g. Java, Python, Go, Rust, ...) with minimum effort to get started with writing the ZIO applications.

The goal of this issue is to gather any number of issues and acceptance criteria for the [documentation project](https://zio.dev/).  A brainstorming session can be held in order to collect more issues that are important to new users. **So please feel free to contribute to this thread, by proposing any change that can help the documentation.**

Here are some of the **unordered** issues (work in progress):

## Content Issues
- Each guide, should have a goal section that instructs the reader on what should be expected after reading the article.
- Make sure quickstarts and tutorials don't use multiple Jargon (except for comprehension, effect, and ZIO). We should try to teach programming using ZIO, only with simple concepts, just by showing the code and how they can do it, without using any special FP jargon.
- Migrate the "Data Types" section to the "Reference" section. This is not just a renaming job. We need to focus on topics instead of data types. So, for example, we should have a separate topic called "State Management" in this topic we introduce the state aspect of the `Ref` data type, while we have another section called "Concurrency" which introduces the concurrent features of the `Ref` data type.
- Having a comprehensive guide on practical topics and quickstart (learning by doing) helps us attract more adopters.
- Writing an article about the [ZIO Petclinic](https://github.com/zio/zio-petclinic/) written by @oliveiosello
- We have missed the FAQ section. This section should address common questions that people might ask, like:
  1. Can I have type classes with ZIO?
  2. Should I learn category theory in order to write ZIO applications?
  3. I'm not a Scala developer and scala has lots of features, what are the minimum levels and topics that I need to learn, in order to start with ZIO?
  4. I work in a team, and our code base is Scala/Java. Can I develop new components using ZIO along with the old components? Is it possible to integrate with the legacy application?
  
   We should gather a list of these questions and answer them on the [faq](https://zio.dev/about/faq).

- Creating another page for defining all programming jargon that we used in the documentation.
- Finding common mistakes when working with functional effects, and documenting them. e.g. Some users wonder why the foo function is not called:
  ```scala
  def performTwoOperation() {
     fooEffect()
     barEffect()
  }
  ```
- The ZIO Streams API has tons of operations, most of the are not memorizable. It is good to add some simple examples for each operation, in the Scala doc.
- The import for zio-streams is not necessary [here](https://zio.dev/getting_started).

## Technical Issues 
We have discussed technical issues and documented them [here](https://github.com/zio/gatsby-theme-zio/blob/main/docs/REQIREMENTS.md). /cc @atooni 

Here are some other issues:
- Integrating [ZIO Start](https://github.com/kitlangton/zio-start) developed by @kitlangton with documentation.
- Integrating the [zio-ecosystem](https://github.com/swoogles/zio-ecosystem) developed by @swoogles with documentation. https://github.com/zio/zio/issues/7037
- Adding analytics to the website to understand user behavior, and which pages and articles are more attractive to users.
- Adding a "Feedback Button" at the end of each page to get feedback from users. This widget can be a very simple question: "Was this page helpful?".

## Updated (25 July)

After talking to @ouertani here are some other issues that we have:
- The _type aliases_ and the way we introduce them can be a source of confusion for new users. 
  - One solution is to introduce them step by step, instead of defining each one as a specific case of the ZIO type.
- Some pages are long to read, and it makes a user hard to find the right content.
  - One solution is to break pages into multiple pages, which we have started for the `ZIO` section.
- The "Error Management" section is hard to find section. Sometimes people find themselves searching for a topic through google instead of finding them through the documentation.
- The contribution guide for documentation has a problem. A user cannot compile and test the documentation, locally.
- The ecosystem section is not under the focus of zio.dev site. So newcomers, can't find the huge ecosystem around the ZIO. 
  - Maybe the idea of having a separate page for "Reference" and "Ecosystem" is not a good idea.
- There is a duplication issue on commented blocks (https://github.com/scalameta/mdoc/issues/321).

## Updated (July 27)
After interviewing with @hnaderi here are other issues:
- Integration example in [the first place](https://zio.dev/getting_started) is extra and confusing. In situations where new users are unaware of safe and unsafe code. 
- The Console example is also a piece of extra information for someone who is new to ZIO. It's better to remove [that section](https://zio.dev/getting_started).
- For someone new to ZIO, the "ZIO application" phrase is weird. What is the ZIO application?
  - It's better to explain to the user, that when you are working with the ZIO library, all of your code should be written as a combination of ZIO workflows. So the ZIO Application is the result of composing ZIO effects.
- In the Introduction pages, there is not something that emphasizes the importance of programming with values, instead of statements.
- providing some showcases "without explaining details" can be very motivational in the first place.
- Comparing ZIO with Future is also something that can help new comments to find out what zio is. 
- Instead of saying what is "functional effect" and what is "ZIO" we can focus on what ZIO isn't?  
- What is "Data Type"? When a Java developer comes and sees that we are calling the "ZIO" as a data type, the developer might compare it to the "POJO"!!!
  - So it is better to explain what we mean by a data type? especially when we explain that the "ZIO" data type is composable.
- The mode of ZIO as `ZEnvironment[R] => Either[Cause[E], A]` is very complex and this first place of documentation. What is `ZEnvironment`? What is `Cause`?
- Explaining ZIO Type Aliases step by step is better than explaining them in one place.
- It is better to introduce ZIO by introducing it to different people. One who comes from the Akka community, one who comes from the cats-effect, who is coming from the Spring, and so on ...
- The `fold` explanation in this [first place](https://zio.dev/getting_started) is not necessary.

## Updated (Aug 02)
- As a user when I'm adopting ZIO I need to know which part of ZIO is macro-based. Can I use ZIO without any macro facilities? 
- As an expert user, I would like to know the internals of ZIO and how it works internally. Knowing these kinds of stuff helps to adopt ZIO easily without hesitation. ZIO Runtime, ZIO STM, ZLayer
- ZLayer is a very new concept that isn't very similar to other dependency injection mechanisms. So it is important for new users to have a clear understanding of ZLayers.
- STM use-cases in real-world applications can help developers to adopt it easily. 
- Introducing the ZIO ecosystem, especially explaining which of them are production ready is necessary.

## Updated (Aug 04)
After talking to @andreabrduque, here are some other issues we should consider:
- For new users, it would be better to have a separate section with a topic on why functional effects are important. e.g. "Why?", "Why Functional Effect?" or "Why ZIO?".
- Current structure is more data-type oriented while focusing on topics is more important. For example, instead of focusing on `Gen` data type, it's better to have a topic on property-based testing.
- Search tool is missing.
- "Service Pattern" section is not accessible easily. fixed in #7194
- Having a section on streaming in the overview section could be useful.
- Having some guides which refer to some topics in the "reference" section would be useful. For example, "how to create streaming API from a callback"
- Adding a badge to each ecosystem library about the status of each project can help users which projects are production ready and we can adopt without a major issue.

**Comment by oliveiosello** [comment]
I think these are all great ideas for improving the documentation, especially the third bullet point regarding restructuring the way concepts are presented. Starting from a higher level and explaining what elements of `ZIO` come into play, rather than presenting from the most specific details and working outward.

Additionally, a FAQ section will definitely be useful, but it might be worth ensuring that the questions we are including have concrete answers. For example, it might be hard to quantify the "amount" of Scala someone would need to know to start learning ZIO. More abstract concepts that require more nuance might be better addressed in a blog post. Because, in this specific example, you can learn both concurrently but it might take some work to convince someone of this. This might apply to questions about mathematical theory, or computer science as well. Because some things, like referential transparency, are worth understanding, but there is a way to explain these concepts that's more approachable than if you were to say "here are the category theory concepts you need to know." Saying no you don't need to know category theory would be a little bit of a lie, and a flat yes might scare people.

Overall, this is a great start in making the documentation easier to consume and understand. I think the right issues are being targeted and the docs will only get better as we move forward.

**Comment by khajavi** [comment]
I agree with what you have said about the FAQ page. The root cause of this problem is the hard-to-learn experience on the tagless final. People often need to learn highly complex topics, type theory, higher kind types, type classes, implicit, and so on to get started with a hello world! So as a marketing phrase, we can say that you can start with ZIO without any deep knowledge of these topics.

**Comment by adamgfraser** [comment]
I think another great use of the FAQ would be getting more content from the Discord into it so that once we answer a question once we have that captured there for others.

**Comment by ghostdogpr** [comment]
I think the FAQ would be good for the questions that get asked on Discord very often and have simple answers such as "forkDaemon" or "ZIO.foreach". Even if people keep asking, we can simply send them a link to the FAQ entry rather than repeating ourselves.

As an example you can check the caliban FAQ https://ghostdogpr.github.io/caliban/faq/ . I tried to make it very visible on the website and feel like it really reduced the amount of "simple" questions.
